src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -50,30 +52,43 @@ def run_python_test(transitive_hydrated_target, pytest):
       for py_req in maybe_python_req_lib.adaptor.requirements:
         all_requirements.append(str(py_req.requirement))
 
-  # TODO: This should be configurable, both with interpreter constraints, and for remote execution.
   # TODO(#7061): This str() can be removed after we drop py2!
   python_binary = text_type(sys.executable)
 
+  # TODO(7680): TargetAdaptors should populate Target attributes with a None value, meaning when
+  # they are not specified in the BUILD file. Here, we need every target adaptor to have
+  # its `.compatibility` defined, so we hackily monkey patch it for now.
```

Comment:
Rather than monkey patching it, would use `getattr(target.adaptor, 'compatibility', None)`. Mutation is scary.

Reply:
Agreed, not a fan of mutation.

This proposal won't work, however. We need the `target.adaptor` to have `compatibility` by the time it gets passed to `python_setup.compatibility_or_constraints`:

https://github.com/pantsbuild/pants/blob/3742edb4040967899c404b43570caa721bb3bd95/src/python/pants/backend/python/subsystems/python_setup.py#L125-L132

We could modify `PythonSetup` to avoid this `AttributeError`, but I'm strongly biased against changing any core `Subsystem` code (this function is used in several important places), especially considering that this workaround is hopefully temporary thanks to #7680.

Reply:
Got it. In that particular case, I would suggest (in descending order):
1. adding another version of the method that takes requirements directly
2. modifying the method to use `getattr`
3. passing in a shim class
4. mutation